### PR TITLE
Fixed #56 -- Add AutoField.is_signed argument

### PR DESCRIPTION
**Pull request by caioariede**
## Ticket

https://code.djangoproject.com/ticket/56

###### Django 3.0 release notes

Screen Shot 2019-10-10 at 9 43 15 AM

###### Model fields: AutoField

Screen Shot 2019-10-10 at 9 42 28 AM

## Timeline

* 2019-10-10 -- PR submission
* 2019-10-15 -- Changed approach - see https://github.com/django/django/pull/11900#issuecomment-542198610
* 2019-10-16 -- Created auxiliary PR #11926 to add `PositiveBigIntegerField` - see https://github.com/django/django/pull/11900#issuecomment-542680058
* 2019-11-19 -- PR #11926 was merged in, now we have `PositiveBigIntegerField`

### CODE REVIEW

**Comment by carltongibson** [comment]
Wow. `Fixed #56 -- ... ` Good work @caioariede! 😀

**Comment by pauloxnet** [comment]
But now if you use **PostgreSQL** as database backend it's impossible to have `True` in `is_signed`  because [AutoField](https://github.com/django/django/blob/master/django/db/backends/postgresql/base.py#L74) use a [serial](https://www.postgresql.org/docs/12/datatype-numeric.html#DATATYPE-NUMERIC-TABLE) that is only positive _(1 to 2147483647)_.

That can be an issue ?

**Comment by caioariede** [comment]
@pauloxnet this should only affect MySQL and SQLite, since the other database backends don't suffer from this issue. For example, [PostgreSQL does not have unsigned data types,](https://stackoverflow.com/questions/20810134/why-unsigned-integer-is-not-available-in-postgresql) and Oracle backend uses `NUMBER` which afaik is already unsigned.

**Comment by pauloxnet** [comment]
@caioariede so `is_signed` has no effect on PostgreSQL and Oracle? With these backends an AutoField will have only positive values?
This different behavior was not clear for me without your clarification, I think can be very helpful for everyone to add this clarification on the `AutoField` documentation section.

**Comment by caioariede** [comment]
@pauloxnet that’s a good point. Indeed a mention in the docs explaining this would be great. Another way for doing this, is perhaps renaming `is_signed` to something like `use_unsigned`, where the documentation could say something like:

> Instructs the database backend to use an unsigned data type when it is supported by the backend

What do you think?

**Comment by pauloxnet** [comment]
> @pauloxnet that’s a good point. Indeed a mention in the docs explaining this would be great. Another way for doing this, is perhaps renaming `is_signed` to something like `use_unsigned`, where the documentation could say something like:
> > Instructs the database backend to use an unsigned data type when it is supported by the backend

I'm not good in rephrasing, and I don't know if would better or more clear in english using `is_signed` or `use_unsigned`.

The main concepts I think we have to put in the `AutoField` documentation are:
- `AutoField` have positive values in every supported database backends
- Only if you use **MySQL** and **SQLlite** you **could optionally** have negative values
- Setting a value in `is_signed` attribute have effects only for **MySQL** and **SQLlite** (e.g. if you set  `is_signed` to `False` in *PostgreSQL* have no effects because [SERIAL](https://www.postgresql.org/docs/12/datatype-numeric.html#DATATYPE-NUMERIC-TABLE) is used anyhow)

Maybe the default value for `is_signed` should be **True** because is the behavior that is valid for every supported database backend ?

BTW I appreciated a lot your work, thanks !!!

**Comment by caioariede** [comment]
> AutoField have positive values in every supported database backends

> Maybe the default value for is_signed should be True because is the behavior that is valid for every supported database backend ?

I guess that's the main goal, not sure if in the long or short-term.

I'm just afraid of changing the default behavior, but since it's a point release I guess we **could** do it. Let's see what the reviewer has to say.

Thanks @pauloxnet :)

**Comment by cansarigol** [comment]
hello good work.  A system check  can be added to avoid confusion maybe.

**Comment by pauloxnet** [comment]
@carltongibson @caioariede do you think that someone have ever used negative values as primary keys ?

I would propose to set unsigned values only for MySQL and SQLite and fix this very old bug.

In that way we will have positive only values in all supported database backends , an theoretically a more portable Django apps.

What do you think ?

**Comment by caioariede** [comment]
I agree with @pauloxnet, I really think we could just go with this as the default. Maybe don’t even have an option to use signed as it would not affect other backends either way

**Comment by carltongibson** [comment]
There may be some use-case. I'm not sure. My _concern_ here would be adding a kwarg that is only useful for a **very** small number of people. If that's the case, it may be a better design overall to not add the extra option. (Thus saving the vast majority those few cycles each time...) 

On the other hand, if it's generally useful, then of course... 

@charettes @felixxm Can I ask your thoughts? Is this worth the extra API surface area?

**Comment by caioariede** [comment]
> My concern here would be adding a kwarg that is only useful for a very small number of people. 

Totally agree. In the other hand, I can see people having issues with large tables where the `id` column used to be `signed` and now will have to migrate to `unsigned`.

A proper solution, could be having `is_signed=False` by default but letting users set it to `True` to avoid the migration, but then emit a warning telling it will be removed in the next release.

What do you guys think?

**Comment by caioariede** [comment]
> Just a note that we strive to avoid gender-oriented terms in the Django ecosystem. Girls are welcome to give their opinion too :-)

Thanks @claudep, I'll update and avoid it in further conversations. Good point!

**Comment by felixxm** [comment]
@caioariede I agree with Carlton then adding a new `is_signed` argument to `AutoField` & co. will make this option totally ignored by ~99.9% of users :man_shrugging: . It is also inconsistent with other field types in Django i.e. we have `IntegerField` and `PositiveIntegerField`, `SmallIntegerField` and `PositiveSmallIntegerField`, etc. (that's why Colin proposed introducing `PositiveAutoField` and `PositiveBigAutoField` in #8183). See also [comment](https://github.com/django/django/pull/49#issuecomment-5666062) about inconsistency between different backends.

In the same time we cannot change a default behavior because it will be backward incompatible and may become an issue for many projects (not only for the big one). It will force all projects to alter all primary keys and foreign keys, what can be extremely tricky :confused: . IMO changing a default behavior is not an option.

It's no coincidence that this ticket is open for 14 years 🤯, and yes I know projects that use negative primary keys :wink: (e.g. for data migrated from legacy databases).

**Comment by carltongibson** [comment]
> In that case, I might be inclined towards `wontfix`...

?

**Comment by caioariede** [comment]
>  IMO changing a default behavior is not an option. 

Even in a point release?

I think a lot of people may not be aware that primary keys are being stored as signed integers. The people using negative pks would probably be in the 99.9% side?

**Comment by caioariede** [comment]
@carltongibson @pauloxnet I've changed the approach here for a solution without the `is_signed` argument, which I personally prefer. It also fixes the consistency issue mentioned by @felixxm :)

Looking forward to your opinions on this.

File: docs/releases/3.0.txt [hunk 1]
```diff
@@ -410,6 +410,10 @@ backends.
   argument, which is an iterable of fields to be returned after insert. Usually
   this is only the auto-generated primary key.
 
+* :class:`~django.db.models.AutoField`, :class:`~django.db.models.BigAutoField`, and
+  :class:`~django.db.models.SmallAutoField` are now defined as unsigned integers on
+  MySQL and SQLite. Historically, they were defined as signed integers.
```
Comment (pauloxnet) [inline docs/releases/3.0.txt]: I would clarify that in PostgreSQL and Oracle they're already unsigned.

**Comment by caioariede** [comment]
@pauloxnet tests are failing due to inconsistency between database fields, as we don't have a consistent relation for `BigAutoField`. I'm currently working on it in #11926 but it's currently a draft.

**Comment by felixxm** [comment]
> Even in a non-point release?

@caioariede We cannot change the current behavior even in the 4.0 release because it would force all SQLite, MySQL, and MariaDB users to migrate all primary keys and foreign keys. On SQLite it's the same as re-creating all tables.

**Comment by caioariede** [comment]
@felixxm cool, let's figure it out. I guess one option would be to have a setting in database options?

I'm gonna put this PR on hold while I work on #11926 but we can keep the discussion.

Thanks

File: docs/releases/3.0.txt (hunk 1 above)
Comment (pope1ni) [reply docs/releases/3.0.txt]: No, this isn't true. In PostgreSQL the sequence will start counting from 1 as Django's auto fields are backed by serial types, but you can happily store negative values manually as the underlying columns are the integer types.

Comment (pope1ni) [reply docs/releases/3.0.txt]: Further, if we're going to have unsigned auto field support for other backends, we should add check constraints as we do for `PositiveIntegerField` in PostgreSQL.

Comment (charettes) [reply docs/releases/3.0.txt]: > In PostgreSQL the sequence will start counting from 1 as Django's auto fields are backed by serial types, but you can happily store negative values manually as the underlying columns are the integer types.

Is this true for identity columns as well?

**Comment by pope1ni** [comment]
I agree with @felixxm. I'd additionally like to point out that some of us actually explicitly set negative values in auto fields for special entries in certain tables.

Unfortunately we'll have to stick with the existing `AutoField` by default for backward compatibility. My comments regarding how we could manage using `BigAutoField` by default in new projects would also apply to then using a new `PositiveBigAutoField` in the case of this ticket: https://github.com/django/django/pull/8924#issuecomment-516792989

File: docs/releases/3.0.txt (hunk 1 above)
Comment (pope1ni) [reply docs/releases/3.0.txt]: It depends on how they were defined:

```sql
$ psql postgres postgres
psql (11.5)
Type "help" for help.

postgres=# create temporary table t1 (id serial primary key);
CREATE TABLE
postgres=# insert into t1 default values;
INSERT 0 1
postgres=# insert into t1 default values;
INSERT 0 1
postgres=# insert into t1 (id) values (-1);
INSERT 0 1
postgres=# select * from t1;
 id 
────
  1
  2
 -1
(3 rows)

postgres=# create temporary table t2 (id integer primary key generated by default as identity);
CREATE TABLE
postgres=# insert into t2 default values;
INSERT 0 1
postgres=# insert into t2 default values;
INSERT 0 1
postgres=# insert into t2 (id) values (-1);
INSERT 0 1
postgres=# insert into t2 (id) overriding user value values (-1);
INSERT 0 1
postgres=# select * from t2;
 id 
────
  1
  2
 -1
  3
(4 rows)

postgres=# create temporary table t3 (id integer primary key generated always as identity);
CREATE TABLE
postgres=# insert into t3 default values;
INSERT 0 1
postgres=# insert into t3 default values;
INSERT 0 1
postgres=# insert into t3 (id) values (-1);
ERROR:  cannot insert into column "id"
DETAIL:  Column "id" is an identity column defined as GENERATED ALWAYS.
HINT:  Use OVERRIDING SYSTEM VALUE to override.
postgres=# insert into t3 (id) overriding system value values (-1);
INSERT 0 1
postgres=# select * from t3;
 id 
────
  1
  2
 -1
(3 rows)
```

**Comment by pBogey** [comment]
Just arrived here from https://code.djangoproject.com/ticket/56. I don't have any experience whatsoever, but for what is worth, I would like having the option to select which type of integer django will use, not only in primary keys, but also when defining OneToOne or ForeignKeys. For example, I am working with a legacy mysql database and I have all models created with inspectdb. I have also added new managed tables that have references to some of the old ones (that I've changed to be managed) and Django created the "table_id" reference as a signed integer. The relationships are only defined in Django logic as far as I can see, as the database relationships haven't changed (no relationships seen in MySQL Workbench). I don't know if this makes sense to you :). So far it seems to be working, but I think it would be good for consistency to be able to choose the same integer types.

**Comment by caioariede** [comment]
@pBogey that sounds interesting, did you submit a ticket for that? I'm pretty sure other folks may have thoughts on this and a new ticket would be the best place to discuss.

Thanks

**Comment by nmfzone** [comment]
Just to notify if the ~0.1% users who use `unsigned` for his primary key is here. 👌
I'm just a little bit surprised Django don't have something like this by default.

**Comment by pBogey** [comment]
> @pBogey that sounds interesting, did you submit a ticket for that? I'm pretty sure other folks may have thoughts on this and a new ticket would be the best place to discuss.
> 
> Thanks

It took me some time, but I think I finally understand :)). The issue is not the type of field defined by the relationships, but the id field itself. Inspect db generated some tables with `AutoField` as id. The issue is that the real tables have unsigned integers as id. So issue remains the with django's default signed integer id field. Waiting for you to merge, I guess :). Good luck, mate.

**Comment by felixxm** [comment]
Closing per ticket.

**Pull request closed without merging**